src/bidiMapper/domains/events/SubscriptionManager.ts

```diff
@@ -158,33 +195,57 @@ export class SubscriptionManager {
     eventMap.set(event, this.#subscriptionPriority++);
   }
 
+  /**
+   * Unsubscribes atomically from all events in the given contexts and channel.
+   */
+  unsubscribeAll(
+    events: Session.SubscribeParametersEvent[],
+    contextIds: (CommonDataTypes.BrowsingContext | null)[],
+    channel: string | null
+  ) {
+    // Assert all contexts are known.
+    for (const contextId of contextIds) {
+      if (contextId !== null) {
+        this.#browsingContextStorage.getKnownContext(contextId);
+      }
+    }
+
+    const eventContextPairs: Array<
+      [
+        eventName: Session.SubscribeParametersEvent,
+        contextId: CommonDataTypes.BrowsingContext | null
+      ]
+    > = cartesianProduct(unrollEvents(events), contextIds);
+
+    // Assert all unsubscriptions are valid.
+    // If any of the unsubscriptions are invalid, do not unsubscribe from anything.
+    eventContextPairs
+      .map(([event, contextId]) =>
+        this.#checkUnsubscribe(event, contextId, channel)
+      )
+      .forEach((unsubscribe) => unsubscribe());
```

Comment:
that is exactly what I meant, when we discussed it

Reply:
Glad to hear that!